guides/keras_hub/mcp_with_keras_hub.py

```diff
@@ -0,0 +1,828 @@
+        that our model sometimes generates, converting it to our standard format.
+
+        Args:
+            response: The raw response from the AI model
+
+        Returns:
+            A list of parsed tool calls
+        """
+        tool_calls = []
+        pattern = r"```tool_code\s*\n([^`]+)\n```"
+        matches = re.findall(pattern, response, re.DOTALL)
+
+        for match in matches:
+            try:
+                tool_call = self._parse_tool_code_call(match.strip())
+                if tool_call:
+                    tool_calls.append(tool_call)
+            except Exception:
+                continue
+
+        return tool_calls
+
+    def _parse_tool_code_call(self, tool_code: str) -> Dict[str, Any]:
+        """
+        Parse a tool_code call into a tool call structure.
+
+        This method converts the function-call format into our standard
+        tool call format with name and arguments.
+
+        Args:
+            tool_code: The tool code string (e.g., "weather.get_weather(city='Tokyo')")
+
+        Returns:
+            A parsed tool call dictionary or None if parsing fails
+        """
+        if "weather.get_weather" in tool_code:
+            city_match = re.search(r'city="([^"]+)"', tool_code)
+            if city_match:
+                return {"name": "weather", "arguments": {"city": city_match.group(1)}}
+        elif "calculator" in tool_code:
+            expression_match = re.search(r'expression="([^"]+)"', tool_code)
+            if expression_match:
+                return {
+                    "name": "calculator",
+                    "arguments": {"expression": expression_match.group(1)},
+                }
+        elif "search." in tool_code:
+            query_match = re.search(r'query="([^"]+)"', tool_code)
+            if query_match:
+                return {"name": "search", "arguments": {"query": query_match.group(1)}}
+
+        return None
+
+    def chat(self, user_input: str) -> str:
+        """
+        Process a user request and return a response.
+
+        This is the main method that orchestrates the entire MCP workflow:
+        1. Builds a prompt with available tools
+        2. Gets a response from the AI model
+        3. Extracts any tool calls from the response
+        4. Executes tools and gets results
+        5. Returns a formatted response to the user
+
+        Args:
+            user_input: The user's request
+
+        Returns:
+            A response that may include tool results or direct AI responses
+        """
+        # Build the prompt with available tools
+        prompt = self._build_prompt(user_input)
+
+        # Get response from the AI model
+        response = self.model.generate(prompt, max_length=512)
+
+        # Extract tool calls from the response
+        tool_calls = self._extract_tool_calls(response)
+
+        if tool_calls:
+            # Safety check: if multiple tool calls found, execute only the first one
```

Comment:
is this for easy demo purpose? if so we should mention why we are only executing one call